zkvm/docs/zkvm-blockchain.md

```diff
@@ -0,0 +1,406 @@
+3. Verify `block.header.previd` equals the
+   [block ID](#block-id)
+   of `prevheader`.
+4. Verify `block.header.timestamp_ms > prevheader.timestamp_ms`.
+5. Verify `block.header.refscount >= 0` and `block.header.refscount <= prevheader.refscount + 1`.
+6. Compute `txroot`, the [Merkle root hash](zkvm-spec.md#merkle-binary-tree) of `block.txs`, using label `"transactions"`.
+7. Verify `txroot == block.header.txroot`.
+8. For each transaction `tx` in block.txs:
+   1. Verify `tx.mintimems == 0` or `tx.mintimems >= block.header.timestamp_ms`.
+   2. Verify `tx.maxtimems == 0` or `tx.maxtimems < block.header.timestamp_ms`.
+   3. If `block.header.version == 1`,
+      verify `tx.version == 1`.
+   4. [Execute](zkvm-spec.md#vm-execution)
+      `tx` to produce transaction log `txlog`.
+   5. Add `txlog` to the list of output logs.
+
+## Apply block
+
+Applying a block causes a node to replace its [blockchain state](#blockchain-state) with the updated state that results.
+
+Inputs:
+- `block`,
+  the block to apply.
+- `state`,
+  the current
+  blockchain state.
+
+Output:
+- New blockchain state.
+
+Procedure:
+1. [Validate](#validate-block)
+   `block` with `prevheader` set to `state.tipheader`.
+2. Let `state′` be `state`.
+3. Remove items from `state′.nonces` where the expiration timestamp is earlier than `block.header.timestamp_ms`.
+4. For each txlog from the validation step,
+   in order:
+   1. [Apply](#apply-transaction-log)
+      the txlog to `state′` to produce `state′′`.
+   2. Set `state′ <- state′′`.
+5. Compute `utxoroot`, the [Merkle patricia root hash](#merkle-patricia-tree) of `state′.utxos` using label `"utxos"`.
+6. Verify `block.header.utxoroot == utxoroot`.
+7. Compute `nonceroot`, the Merkle patricia root hash of `state′.nonces` using label `"nonces"`.
+8. Verify `block.header.nonceroot == nonceroot`.
+9. Set `state′.tipheader <- block.header`.
+10. Add `block.header` to the end of the `state′.refids` list.
+11. Prune `state′.refids` to the number of items specified by `block.header.refscount` by removing the oldest IDs.
+12. Return `state′`.
+
+## Apply transaction log
+
+Inputs:
+- `txlog`,
+  a
+  [transaction log](zkvm-spec.md#transaction-log).
+- `state`,
+  a
+  [blockchain state](#blockchain-state).
+
+Output:
+- New blockchain state.
+
+Procedure:
+1. Let `state′` be `state`.
+2. For each
+   [nonce entry](zkvm-spec.md#nonce-entry)
+   `n` in `txlog`:
+   1. Verify `n.blockid` is one of the following:
+      - The [ID](#block-id) of `state.initialheader`,
+        or
+      - One of the block ids in `state.refids`.
+   2. Verify `n` is _not_ in `state.nonces`.
+   3. Add `n` to `state′.nonces`.
+3. For each
+   [input entry](zkvm-spec.md#input-entry)
+   or
+   [output entry](zkvm-spec.md#output-entry)
+   in `txlog`:
+   1. If an input entry,
+      verify its ID is in `state′.utxos`,
+      then remove it.
+   2. If an output entry,
+      add its utxo ID to `state′.utxos`.
```

Comment:
Maybe add a note that added utxos may be immediately consumed in the _same transaction_ or in the _same block_. So no one tries to remove and add in bulk, without annihilating such local pairs first.